edk2toollib/uefi/telem/parser_main.py

```diff
@@ -0,0 +1,660 @@
+I           unsigned int            integer         4
+l           long                    integer         4
+L           unsigned long           integer         4
+q           long long               integer         8
+Q           unsigned long long      integer         8
+n           ssize_t                 integer
+N           size_t                  integer
+e           (6)                     float           2
+f           float                   float           4
+d           double                  float           8
+s           char[]                  bytes
+p           char[]                  bytes
+P           void *                  integer
+"""
+
+CPER_HEADER_SIZE            = 128 # A CPER header is 128 bytes 
+CPER_SECTION_HEADER_SIZE    = 72  # A CPER section header is 72 bytes
+FriendlyNames               = {}  # dict loaded from friendlynames.csv used to associate guids with friendly strings
+Parsers                     = []  # list of plugin classes which inherit from CPER_SECTION_DATA and are therefore capable of parsing section data
+
+# Using these lambda concat statements to guarantee linear runtime when concatonating strings, and for simplicity
+concat_s = lambda x,y  : (''.join([x,y," "]))  # Concatenate two strings and put a space at the end
+concat_n = lambda x,y  : (''.join([x,y,"\n"])) # Concatenate two strings and put a newline at the end
+concat   = lambda x,y  : (''.join([x,y]))      # Concatenate two strings
+
+# Useful lambdas
+empty    = lambda x     : x == "" or x == None or x == []   # True if x is an empty string, list, etc.
+neq      = lambda x,y   : x != y                            # True if x and y are not equal
+eq       = lambda x,y   : x == y                            # True if x and y are equal
+
+class CPER(object):
+
+    def __init__(self, input):
+        self.RawData = bytearray.fromhex(input)
+        self.Header = None
+        self.SetCPERHeader()
+        self.SectionHeaders = []
+        self.SetSectionHeaders()
+
+    ##
+    # Turn the portion of the raw input associated with the CPER head into a CPER_HEAD object
+    ##
+    def SetCPERHeader(self):
+        temp = self.RawData[:CPER_HEADER_SIZE]
+        self.Header = CPER_HEADER(temp)
+
+    ##
+    # Set each of the section headers to CPER_SECTION_HEADER objects
+    ##
+    def SetSectionHeaders(self):
+        # Section of RawData containing the section headers
+        temp = self.RawData[CPER_HEADER_SIZE:CPER_HEADER_SIZE + (CPER_SECTION_HEADER_SIZE * self.Header.SectionCount)]
+
+        # Go through each section header and attempt to create a CPER_SECTION_HEADER object. Store each header in SectionHeaders list
+        for x in range(self.Header.SectionCount):
+            try: # Don't want to stop runtime if parsing fails
+                self.SectionHeaders.append(CPER_SECTION_HEADER(temp[x * CPER_SECTION_HEADER_SIZE: (x + 1) * CPER_SECTION_HEADER_SIZE]))
+            except:
```

Comment:
Try to put specific exception types here. Like array out of bounds would be the main one